src/neural/cuda/layers.cc

```diff
@@ -1225,6 +1141,424 @@ ResidualBlock<DataType>::~ResidualBlock() {
+
+  }  // End of encoder blocks
+
+  // queries (policy/attention/wq) -> scratch 1
+  {
+    const int num_inputs = embedding_op_size_;
+    const int num_outputs = policy_d_model_;
+    const int batch = N * 64;
+    ReportCUBLASErrors(cublasHgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N,
+                                   num_outputs, batch, num_inputs, &alpha,
+                                   ip2_pol_w_, num_inputs, scratch0, num_inputs,
+                                   &beta, scratch1, num_outputs));
+    addVectors(scratch1, ip2_pol_b_, scratch1, num_outputs * batch, num_outputs,
+               num_outputs * batch, NONE, stream);
+  }
+
+  // keys (policy/attention/wk) -> scratch 2
+  {
+    const int num_inputs = embedding_op_size_;
+    const int num_outputs = policy_d_model_;
+    const int batch = N * 64;
+    ReportCUBLASErrors(cublasHgemm(cublas, CUBLAS_OP_T, CUBLAS_OP_N,
+                                   num_outputs, batch, num_inputs, &alpha,
+                                   ip3_pol_w_, num_inputs, scratch0, num_inputs,
+                                   &beta, scratch2, num_outputs));
+    addVectors(scratch2, ip3_pol_b_, scratch2, num_outputs * batch, num_outputs,
+               num_outputs * batch, NONE, stream);
+  }
+
+  // dk = tf.math.sqrt(tf.cast(tf.shape(keys)[-1], self.model_dtype))  # constant for scaling
+  // # POLICY SELF-ATTENTION: self-attention weights are interpreted as from->to policy
+  // matmul_qk = tf.matmul(queries, keys, transpose_b=True)  # Bx64x64 (from 64 queries, 64 keys)
+  // policy_attn_logits = matmul_qk / dk       # Bx64x64 (64 from-squares, 64 to-squares)
+  {
+    // shape(keys)[-1] = policy_d_model_
+    unsigned short factor = float_to_half(1.0f / sqrt((float)policy_d_model_));
+    cublasGemmStridedBatchedEx(cublas, CUBLAS_OP_T, CUBLAS_OP_N, 
+                               64 /*M*/, 64 /*N*/, policy_d_model_ /*K*/,    // A/B, and M/N are swapped for row-major to col-major transform
+                               &factor,                            // to handle "/ tf.math.sqrt(dk)"
+                               scratch2 /*A*/,
+                               CUDA_R_16F, 
+                               policy_d_model_ /*LDA*/,
+                               64 * policy_d_model_,    /*strideA*/
+                               scratch1 /*B*/, 
+                               CUDA_R_16F, 
+                               policy_d_model_ /*LDB*/,
+                               64 * policy_d_model_,     /*strideB*/
+                               &beta, 
+                               output /*C*/,  // output (policy_attn_logits)
+                               CUDA_R_16F, 
+                               64 /*LDC*/, 
+                               64 * 64 /*strideC*/,
+                               N, CUDA_R_16F, CUBLAS_GEMM_DEFAULT);
+  }
+
+  // Compute promotion_logits in a single kernel (and put the result just after policy_attn_logits to get concat for free)
```

Comment:
This logic seems wrong - you don't get the concat for free since policy map is expecting an input in format N *(64*64+8*24), not N * 64* 64 + N*8*24)
I think you need to set strideC to 64*64+8*24 and change the ComputePromotionLogits implementation to interleave the output as appropriate.
As it stands this will work for N=1 only I think...